storage/src/merkle_storage.rs

```diff
@@ -877,177 +757,85 @@ impl MerkleStorage {
             message,
         };
         let entry = Entry::Commit(new_commit.clone());
-
-        let new_commit_hash = hash_commit(&new_commit)?;
-        self.put_to_staging_area(&new_commit_hash, entry.clone())?;
+        self.put_to_staging_area(&hash_commit(&new_commit), entry.clone());
         self.persist_staged_entry_to_db(&entry)?;
-        self.staged = Vec::new();
-        self.staged_indices = HashMap::new();
-        self.last_commit_hash = Some(new_commit_hash.clone());
 
-        self.update_execution_stats("Commit".to_string(), None, &instant);
-        Ok(new_commit_hash)
+        self.staged.clear();
+        self.last_commit_hash = Some(hash_commit(&new_commit.clone()));
+        Ok(hash_commit(&new_commit))
     }
 
-    pub fn start_new_cycle(&mut self) -> Result<(), MerkleError> {
-        let instant = Instant::now();
-        self.db.start_new_cycle(self.last_commit_hash.clone());
-        self.update_execution_stats("GC".to_string(), None, &instant);
-        Ok(())
+    /// Set key/val to the staging area.
+    fn set_stage_root(&mut self, tree: &Tree) {
+        self.current_stage_tree = (tree.clone(), hash_tree(&tree));
     }
 
     /// Set key/val to the staging area.
     pub fn set(&mut self, key: &ContextKey, value: &ContextValue) -> Result<(), MerkleError> {
-        let instant = Instant::now();
-        let act = Arc::make_mut(&mut self.actions);
-        // store action
-        act.push(Action::Set(SetAction {
-            key: key.to_vec(),
-            value: value.to_vec(),
-        }));
-        self.update_execution_stats("Set".to_string(), Some(&key), &instant);
+        let _ = StatUpdater::new(& mut self.stats, MerkleStorageAction::Set, Some(key));
+        let root = self.get_staged_root();
+        let new_root_hash = &self._set(&root, key, value)?;
+        self.set_stage_root(&self.get_tree(new_root_hash)?);
         Ok(())
     }
 
+    fn _set(
+        &mut self,
+        root: &Tree,
+        key: &ContextKey,
+        value: &ContextValue,
+    ) -> Result<EntryHash, MerkleError> {
+        let blob_hash = hash_blob(&value);
+        self.put_to_staging_area(&blob_hash, Entry::Blob(value.clone()));
+        let new_node = Node {
+            entry_hash: blob_hash,
+            node_kind: NodeKind::Leaf,
+        };
+        let rv = self.compute_new_root_with_change(root, &key, Some(new_node));
```

Comment:
Other than for being able to play well with the current way actions are recorded, do we need to calculate the hash every time? cc @bkontur

Reply:
that gonna be optimized a little bit later ;)